### PR TITLE
add a mongoengine model for repositories

### PR DESCRIPTION
**Pull request by asmacdo**
closes: [#785](https://pulp.plan.io/issues/785)

### CODE REVIEW

File: server/pulp/server/db/model/__init__.py [hunk 1]
```diff
@@ -51,3 +52,51 @@ class RepositoryContentUnit(Document):
...
+    :ivar notes: arbitrary key-value pairs programmatically describing the repo;
+                 these are intended as a way to describe the repo usage or
+                 organizational purposes and should not vary depending on the
+                 actual content of the repo
+    :type notes: mongoengine.DictField
+    :ivar content_unit_counts: key-value pairs of number of units associated with this repo.
+                               This is different than the number of associations, since a
+                               unit may be associated multiple times.
+    :type content_unit_counts: mongoengine.DictField
+    :ivar metadata: arbitrary data that describes the contents of the repo;
```
Comment (ipanova) [inline server/pulp/server/db/model/__init__.py]: why do you have the metadata described here? i do not see it among the attributes

File: server/pulp/server/db/model/__init__.py [hunk 2]
```diff
@@ -51,3 +52,51 @@ class RepositoryContentUnit(Document):
...
+    # obj._id. So we name this field `repo_id`, and temporarily continue to store it in the db as
+    # `id` for backwards compatibility. This should be migrated to `repo_id` in a future X release.
+    repo_id = StringField(db_field='id', required=True)
+    display_name = StringField(required=True)
+    description = StringField()
+    notes = DictField()
+    scratchpad = DictField()
+    content_unit_counts = DictField()
+    last_unit_added = DateTimeField()
+    last_unit_removed = DateTimeField()
```
Comment (ipanova) [inline server/pulp/server/db/model/__init__.py]: please add `last_unit_added` and `last_unit_removed` description to the doc block

File: server/pulp/server/db/model/__init__.py [hunk 3]
```diff
@@ -51,3 +52,51 @@ class RepositoryContentUnit(Document):
...
+    :ivar _ns: (Deprecated) Namespace of repo, included for backwards compatibility.
+    :type _is: mongoengine.StringField
+    """
+
+    # This field is named `repo_id` because `id`  cannot be accessed using the normal mongoengine
+    # idiom of `obj.id` because `obj.id` has been aliased (internally in mongoengine) to reference
+    # obj._id. So we name this field `repo_id`, and temporarily continue to store it in the db as
+    # `id` for backwards compatibility. This should be migrated to `repo_id` in a future X release.
+    repo_id = StringField(db_field='id', required=True)
+    display_name = StringField(required=True)
```
Comment (ipanova) [inline server/pulp/server/db/model/__init__.py]: can you explain why have you decided to make display_name as required field and, for example, description not required?

File: server/pulp/server/db/model/__init__.py [hunk 4]
```diff
@@ -51,3 +52,51 @@ class RepositoryContentUnit(Document):
...
+    description = StringField()
+    notes = DictField()
+    scratchpad = DictField()
```
Comment (ipanova) [inline server/pulp/server/db/model/__init__.py]: please add `scratchpad` description to the doc block

File: server/pulp/server/db/model/__init__.py [hunk 5]
```diff
@@ -51,3 +52,51 @@ class RepositoryContentUnit(Document):
...
+    content_unit_counts = DictField()
+    last_unit_added = DateTimeField()
```
Comment (ipanova) [inline server/pulp/server/db/model/__init__.py]: I am not sure, but should not we use the ISO8601StringField here?

File: server/pulp/server/db/manage.py [hunk 6]
```diff
@@ -116,6 +116,7 @@ def ensure_database_indexes():
     collection does not already exist.
     """
     model.RepositoryContentUnit.ensure_indexes()
+    model.Repository.ensure_indexes()
```
Comment (ipanova) [inline server/pulp/server/db/manage.py]: please add the testcase for this like here https://github.com/pulp/pulp/blob/master/server/test/unit/server/db/test_manage.py#L118

**Comment by ipanova** [comment]
@asmacdo  assigning this back to you, ping me when you're ready

File: server/pulp/server/db/model/__init__.py (hunk 1 above)
Comment (asmacdo) [inline server/pulp/server/db/model/__init__.py]: I copied this over from the Repo "model". I guess this was removed at some point. Good catch.

File: server/pulp/server/db/model/__init__.py (hunk 3 above)
Comment (asmacdo) [inline server/pulp/server/db/model/__init__.py]: When creating a repo, if `display_name` was not specified, it is automatically set to `repo_id`, meaning there can never be a repo without a `display_name`. However, if `description` is not specified, it remains null.

File: server/pulp/server/db/model/__init__.py (hunk 5 above)
Comment (asmacdo) [inline server/pulp/server/db/model/__init__.py]: the ISO8601StringField is actually not ideal, but is necessary for backwards compatibility. It stores a string equivalent of a real DateTime object. In this place, we are able to actually use a proper DateTime object, so DateTimeField is correct.

**Comment by asmacdo** [comment]
ok test

**Comment by asmacdo** [comment]
@ipanova, changes made, assigned back to you.

File: server/test/unit/server/db/test_manage.py [hunk 7]
```diff
@@ -116,6 +116,7 @@ def test_ensure_index(self, mock_model):
         self.assertTrue(mock_model.ReservedResource.ensure_indexes.called)
         self.assertTrue(mock_model.TaskStatus.ensure_indexes.called)
         self.assertTrue(mock_model.Worker.ensure_indexes.called)
+        self.assertTrue(mock_model.Repository.ensure_indexes.called)
```
Comment (ipanova) [inline server/test/unit/server/db/test_manage.py]: I'd recommend make it in alphabetical order.

File: server/test/unit/server/db/test_model.py [hunk 8]
```diff
@@ -152,3 +152,67 @@ def test_meta_collection(self):
...
+        Ensure that the Repository model is a subclass of Mongoengine's Document class.
+        """
+        sample_model = model.Repository()
+        self.assertTrue(isinstance(sample_model, Document))
+
+    def test_model_fields(self):
+        """
+        Assert that each field has the correct type and `required` status.
+        """
+        self.assertTrue(isinstance(model.Repository.repo_id, StringField))
```
Comment (ipanova) [inline server/test/unit/server/db/test_model.py]: you could also check repo_id.db_field

**Comment by ipanova** [comment]
@asmacdo There are only two small comments. LGTM after you'll make the changes. Well done ^_^

File: server/test/unit/server/db/test_model.py (hunk 8 above)
Comment (asmacdo) [inline server/test/unit/server/db/test_model.py]: good call.

**Comment by asmacdo** [comment]
ok test

**Pull request merged**